### PR TITLE
Synchronous python CDK: make sync mode agnostic

### PR DESCRIPTION
**Pull request by erohmensing**
Always checkpoint state after the slice. Definitely cleanup needed

don't emit sentinel state in full refresh mode

add note when checking that we can't remove the sentinel for the no slices case

don't put checkpoint intervals under incremental only

### CODE REVIEW

**Comment by erohmensing** [comment]
* **#36999** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/36999?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @erohmensing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py [hunk 1]
```diff
@@ -35,6 +35,11 @@
 # value is used to indicate that stream should not load the incoming state value
 FULL_REFRESH_SENTINEL_STATE_KEY = "__ab_full_refresh_state_message"
 
+# Streams that don't have cursors should emit this
+# We still want to emit a state message at the end of the stream
+# So that we can return record counts, but we aren't returning a cursor
+NO_CURSOR_SENTINEL_STATE_KEY = "__ab_no_cursor_state_message"
```
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:41]: Just a heads up that as the temporary workaround we skipped setting state when we saw the old sentinel value here:
https://github.com/airbytehq/airbyte/blob/e74d9363503c6672723ee8db452df824f51e0d9d/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py#L221-L223

Once the platform starts deciding on when to pass state then that logic can get removed and we always trust incoming state

Comment (erohmensing) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/core.py]: heck yeah! [59916f9](https://github.com/airbytehq/airbyte/pull/36999/commits/59916f9c0025d165374407860fcd142714dd63f4)

**Comment by brianjlai** [comment]
@erohmensing gonna close this one out since this should mostly be covered by the changes when I merged https://github.com/airbytehq/airbyte/pull/37429 . Feel free to reopen if you think we still need to address this

**Pull request closed without merging**